packages/flutter_tools/lib/src/android/gradle_utils.dart

```diff
@@ -145,65 +171,426 @@ String getGradleVersionForAndroidPlugin(Directory directory, Logger logger) {
+    max: '3.3.3',
+  )) {
+    return isWithinVersionRange(gradleV,
+        min: '4.10.1', max: _maxKnownAndSupportedGradleVersion);
+  }
+
+  logger.printTrace('Unknown Gradle-Agp compatibility, $gradleV, $agpV');
+  return false;
+}
+
+// Validate that the [javaVersion] and Gradle version are compatible with
+// each other.
+//
+// Source of truth:
+// https://docs.gradle.org/current/userguide/compatibility.html#java
+bool validateJavaGradle(Logger logger,
+    {required String? javaV, required String? gradleV}) {
+  // Update these when new major versions of Java are supported by android.
+  // Supported means Java <-> Gradle support.
+  const String oneMajorVersionHigherJavaVersion = '20';
+
+  // https://docs.gradle.org/current/userguide/compatibility.html#java
+  const String oldestSupportedJavaVersion = '1.8';
+  const String oldestDocumentedJavaGradleCompatibility = '2.0';
+
+  // Begin Java <-> Gradle validation.
+
+  if (javaV == null || gradleV == null) {
+    logger.printTrace(
+        'Java version or Gradle version unknown ($javaV, $gradleV).');
+    return false;
   }
-  if (_isWithinVersionRange(androidPluginVersion, min: '3.1.0', max: '3.1.9')) {
-    return '4.4';
+
+  // First check if versions are too old.
+  if (isWithinVersionRange(javaV,
+      min: '1.1', max: oldestSupportedJavaVersion, inclusiveMax: false)) {
+    logger.printTrace('Java Version: $javaV is too old.');
+    return false;
   }
-  if (_isWithinVersionRange(androidPluginVersion, min: '3.2.0', max: '3.2.1')) {
-    return '4.6';
+  if (isWithinVersionRange(gradleV,
+      min: '0.0', max: oldestDocumentedJavaGradleCompatibility, inclusiveMax: false)) {
+    logger.printTrace('Gradle Version: $gradleV is too old.');
+    return false;
   }
-  if (_isWithinVersionRange(androidPluginVersion, min: '3.3.0', max: '3.3.2')) {
-    return '4.10.2';
+
+  // Check if versions are newer than the max supported versions.
+  if (isWithinVersionRange(
+    javaV,
+    min: oneMajorVersionHigherJavaVersion,
+    max: '100.100',
+  )) {
+    // Assume versions Java versions newer than [maxSupportedJavaVersion]
+    // required a higher gradle version.
+    final bool validGradle = isWithinVersionRange(gradleV,
+        min: _maxKnownAndSupportedGradleVersion, max: '100.00');
+    logger.printWarning(
+        'Newer than known valid Java version ($javaV), gradle ($gradleV).'
+        '\n Treating as valid configuration.');
+    return validGradle;
   }
-  if (_isWithinVersionRange(androidPluginVersion, min: '3.4.0', max: '3.5.0')) {
-    return '5.6.2';
+
+  // Begin known Java <-> Gradle evaluation.
+    final List<JavaGradleCompat> compatList = <JavaGradleCompat>[
+    JavaGradleCompat(
+      javaMin: '19',
+      javaMax: '20',
+      gradleMin: '7.6',
+      gradleMax: _maxKnownAndSupportedGradleVersion,
+    ),
+    JavaGradleCompat(
+      javaMin: '18',
+      javaMax: '19',
+      gradleMin: '7.5',
+      gradleMax: _maxKnownAndSupportedGradleVersion,
+    ),
```

Comment:
I think it would be reasonable to give `gradleMax` a default value in the constructor here. Or to have the loop body say `max: _maxKnownAndSupportedGradleVersion` instead of `max: data.gradleMax`, and just not have a `gradleMax` field in the class.

Then this list can get a lot more compact, which I think will make it easier to scan through and spot patterns or possible problems in:
```suggestion
    JavaGradleCompat(javaMin: '19', javaMax: '20', gradleMin: '7.6'),
    JavaGradleCompat(javaMin: '18', javaMax: '19', gradleMin: '7.5'),
```